src/components/BulkMemberManager/hooks/useResults/useResults.js

```diff
@@ -0,0 +1,108 @@
+import { useState, useEffect, useRef } from 'react'
+import { useDebounce } from 'use-debounce'
+import { useSet } from './useSet'
+
+export const useResults = ({
+    membersQuery,
+    nonMembersQuery,
+    transformQueryResponse,
+    filterDebounceMs,
+    mode,
+}) => {
+    const membersDataQuery = useDataQuery(membersQuery, { lazy: true })
+    const nonMembersDataQuery = useDataQuery(nonMembersQuery, { lazy: true })
+    const [prevMembers, setPrevMembers] = useState()
+    const [prevNonMembers, setPrevNonMembers] = useState()
+    const [membersPage, setMembersPage] = useState(1)
+    const [nonMembersPage, setNonMembersPage] = useState(1)
+    const selectedMembers = useSet()
+    const selectedNonMembers = useSet()
+    const [membersFilter, setMembersFilter] = useState('')
+    const [nonMembersFilter, setNonMembersFilter] = useState('')
+    const [debouncedMembersFilter] = useDebounce(
+        membersFilter,
+        filterDebounceMs
+    )
+    const [debouncedNonMembersFilter] = useDebounce(
+        nonMembersFilter,
+        filterDebounceMs
+    )
+    const transformQueryResponseRef = useRef()
+    transformQueryResponseRef.current = transformQueryResponse
+
+    useEffect(() => {
+        if (mode === 'MEMBERS') {
+            setPrevMembers(
+                membersDataQuery.data &&
+                    transformQueryResponseRef.current(
+                        membersDataQuery.data.results
+                    )
+            )
+            membersDataQuery.refetch({
+                page: membersPage,
+                filter: debouncedMembersFilter,
+            })
+        } else {
+            setPrevNonMembers(
+                nonMembersDataQuery.data &&
+                    transformQueryResponseRef.current(
+                        nonMembersDataQuery.data.results
+                    )
+            )
+            nonMembersDataQuery.refetch({
+                page: nonMembersPage,
+                filter: debouncedNonMembersFilter,
+            })
+        }
+    }, [
+        mode,
+        membersPage,
+        nonMembersPage,
+        debouncedMembersFilter,
+        debouncedNonMembersFilter,
+    ])
+
+    const { called, loading, error, data } =
+        mode === 'MEMBERS' ? membersDataQuery : nonMembersDataQuery
+    const prevResults = mode === 'MEMBERS' ? prevMembers : prevNonMembers
+    const results = data ? transformQueryResponse(data.results) : prevResults
+    const [page, setPage] =
+        mode === 'MEMBERS'
+            ? [membersPage, setMembersPage]
+            : [nonMembersPage, setNonMembersPage]
+    const selected = mode === 'MEMBERS' ? selectedMembers : selectedNonMembers
+    const [filter, setFilter] =
+        mode === 'MEMBERS'
+            ? [membersFilter, setMembersFilter]
+            : [nonMembersFilter, setNonMembersFilter]
+
+    return {
+        loading: !called || loading,
+        error,
+        results,
+        pager: {
```

Comment:
`BulkMemberManager` assumes that the queries provided return paginated data. I think this is a fair assumption but maybe we should have an assertion somewhere that this is the case? That could allow us to provide a better error message if the consumer of this component does not realise that they passed unpaginated queries?